### PR TITLE
[video_player] [iOS] Fixed a memory leak in the video player plugin.

### PR DESCRIPTION
**Pull request by shihchanghsiungsonos**
## Description

This fixes a memory leak in iOS.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. [shared_preferences]
- [ ] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [ ] I updated CHANGELOG.md to add a description of the change.
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://www.dartlang.org/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

File: packages/video_player/ios/Classes/VideoPlayerPlugin.m [hunk 1]
```diff
@@ -82,14 +83,13 @@ - (void)addObservers:(AVPlayerItem*)item {
             options:NSKeyValueObservingOptionInitial | NSKeyValueObservingOptionNew
             context:playbackBufferFullContext];
 
-  [[NSNotificationCenter defaultCenter] addObserverForName:AVPlayerItemDidPlayToEndTimeNotification
+  _notificationObserverId = [[NSNotificationCenter defaultCenter] addObserverForName:AVPlayerItemDidPlayToEndTimeNotification
```
Comment (cyanglaz) [inline packages/video_player/ios/Classes/VideoPlayerPlugin.m]: It might be cleaner to just use the `addObserver:self` here since we are just using mainQueue. This way we can separate the observing implementation out of the `addObservers` method.

File: packages/video_player/ios/Classes/VideoPlayerPlugin.m [hunk 2]
```diff
@@ -395,7 +395,7 @@ - (void)dispose {
                              forKeyPath:@"playbackBufferFull"
                                 context:playbackBufferFullContext];
   [_player replaceCurrentItemWithPlayerItem:nil];
-  [[NSNotificationCenter defaultCenter] removeObserver:self];
+  [[NSNotificationCenter defaultCenter] removeObserver:_notificationObserverId];
```
Comment (cyanglaz) [inline packages/video_player/ios/Classes/VideoPlayerPlugin.m]: This can be unchanged if we use the `addObserver:` method as I mentioned in the other comment.

File: packages/video_player/ios/Classes/VideoPlayerPlugin.m [hunk 3]
```diff
@@ -82,14 +83,13 @@ - (void)addObservers:(AVPlayerItem*)item {
...
-  [[NSNotificationCenter defaultCenter] addObserverForName:AVPlayerItemDidPlayToEndTimeNotification
+  _notificationObserverId = [[NSNotificationCenter defaultCenter] addObserverForName:AVPlayerItemDidPlayToEndTimeNotification
                                                     object:[_player currentItem]
                                                      queue:[NSOperationQueue mainQueue]
                                                 usingBlock:^(NSNotification* note) {
                                                   if (self->_isLooping) {
                                                     AVPlayerItem* p = [note object];
-                                                    [p seekToTime:kCMTimeZero
-                                                        completionHandler:nil];
+                                                    [p seekToTime:kCMTimeZero];
```
Comment (cyanglaz) [inline packages/video_player/ios/Classes/VideoPlayerPlugin.m]: It seems not related.

**Comment by shihchanghsiungsonos** [comment]
@cyanglaz Thanks for the feedback. We will update the PR.

File: packages/video_player/ios/Classes/VideoPlayerPlugin.m [hunk 4]
```diff
@@ -82,20 +82,23 @@ - (void)addObservers:(AVPlayerItem*)item {
...
-                                                  }
-                                                }];
+  // Add an observer that will respond to itemDidPlayToEndTime
+  [[NSNotificationCenter defaultCenter] addObserver:self
+                                           selector:@selector(itemDidPlayToEndTime:)
+                                               name:AVPlayerItemDidPlayToEndTimeNotification
+                                             object:item];
+}
+
+- (void)itemDidPlayToEndTime:(NSNotification*)note {
```
Comment (cyanglaz) [inline packages/video_player/ios/Classes/VideoPlayerPlugin.m]: nits: renaming note -> notification

File: packages/video_player/ios/Classes/VideoPlayerPlugin.m [hunk 5]
```diff
@@ -82,20 +82,23 @@ - (void)addObservers:(AVPlayerItem*)item {
...
+  AVPlayerItem* p = [note object];
+  if (self->_isLooping) {
```
Comment (cyanglaz) [inline packages/video_player/ios/Classes/VideoPlayerPlugin.m]: nits: Same as below for `_eventSink`
```suggestion
  if (_isLooping) {
```

File: packages/video_player/CHANGELOG.md [hunk 6]
```diff
@@ -1,3 +1,7 @@
+## 0.10.1+5
+
+* [iOS] Fixed a memory leak in the video player plugin.
```
Comment (cyanglaz) [inline packages/video_player/CHANGELOG.md]:
```suggestion
* [iOS] Fixed a memory leak with notification observing.
```

File: packages/video_player/ios/Classes/VideoPlayerPlugin.m (hunk 4 above)
Comment (shihchanghsiungsonos) [reply packages/video_player/ios/Classes/VideoPlayerPlugin.m]: Sure.

File: packages/video_player/ios/Classes/VideoPlayerPlugin.m (hunk 5 above)
Comment (shihchanghsiungsonos) [reply packages/video_player/ios/Classes/VideoPlayerPlugin.m]: Will update. Thanks for the feedback!
